File: scripts/scan-all-skills.py
#!/usr/bin/env python3
"""Scan all projects for .claude/skills folders and list skills."""
import os
import json
from pathlib import Path

SCAN_DIRS = [
    Path(r"C:\Users\julien\OneDrive\Coding\_Projets de code"),
    Path(r"C:\Users\julien\OneDrive\Coding\_référentiels de code"),
]
EXCLUDE = ["2025.11 Claude Code MarketPlace"]

def find_skills(base_path: Path) -> dict:
    """Find all skills in a directory tree."""
    results = {}

    for item in base_path.iterdir():
        if not item.is_dir():
            continue
        if item.name in EXCLUDE:
            continue
        if item.name.startswith('.'):
            continue

        skills_path = item / ".claude" / "skills"
        if skills_path.exists() and skills_path.is_dir():
            skills = []
            for skill_dir in skills_path.iterdir():
                if skill_dir.is_dir():
                    skill_md = skill_dir / "SKILL.md"
                    if skill_md.exists():
                        skills.append({
                            "name": skill_dir.name,
                            "type": "skill",
                            "path": str(skill_dir)
                        })
                    else:
                        skills.append({
                            "name": skill_dir.name,
                            "type": "dir_no_skill",
                            "path": str(skill_dir)
                        })
                elif skill_dir.is_file() and skill_dir.suffix == ".md":
                    skills.append({
                        "name": skill_dir.name,
                        "type": "file",
                        "path": str(skill_dir)
                    })

            if skills:
                results[item.name] = {
                    "path": str(item),
                    "skills_count": len([s for s in skills if s["type"] == "skill"]),
                    "items": skills
                }

    return results

def main():
    all_results = {}
    total_skills = 0

    for scan_dir in SCAN_DIRS:
        if scan_dir.exists():
            results = find_skills(scan_dir)
            all_results[str(scan_dir)] = results
            for project, data in results.items():
                total_skills += data["skills_count"]

    # Print summary
    print(f"=== SCAN RESULTS ===\n")
    print(f"Total skills found: {total_skills}\n")

    for scan_dir, projects in all_results.items():
        if projects:
            print(f"\n--- {scan_dir} ---\n")
            for project, data in sorted(projects.items()):
                skill_count = data["skills_count"]
                other_count = len(data["items"]) - skill_count
                print(f"{project}: {skill_count} skills", end="")
                if other_count > 0:
                    print(f" (+{other_count} other files)", end="")
                print()
                for item in data["items"]:
                    marker = "[SKILL]" if item["type"] == "skill" else "[FILE]" if item["type"] == "file" else "[DIR]"
                    print(f"    {marker} {item['name']}")

    # Save JSON
    output_path = Path(__file__).parent.parent / "docs" / "skills-scan-results.json"
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump({
            "total_skills": total_skills,
            "results": all_results
        }, f, indent=2, ensure_ascii=False)
    print(f"\n\nJSON saved to: {output_path}")

if __name__ == "__main__":
    main()

File: scripts/discovery/scan-all-skills.js
#!/usr/bin/env node
/**
 * Scan all projects for .claude/skills folders and list skills.
 *
 * Scans predefined directories for projects containing skills,
 * categorizes items as: skill (with SKILL.md), dir_no_skill, or file (.md files).
 *
 * Pattern from scan-all-skills.py
 */

const fs = require('fs');
const path = require('path');
const { writeJson } = require('../lib/file-utils');

// Directories to scan
const SCAN_DIRS = [
    'C:\\Users\\julien\\OneDrive\\Coding\\_Projets de code',
    'C:\\Users\\julien\\OneDrive\\Coding\\_référentiels de code'
];

// Projects to exclude
const EXCLUDE = ['2025.11 Claude Code MarketPlace'];

/**
 * Find all skills in a directory tree.
 *
 * @param {string} basePath - Base directory to scan
 * @returns {Object} Results by project name
 */
function findSkills(basePath) {
    const results = {};

    // Check if base path exists
    if (!fs.existsSync(basePath)) {
        return results;
    }

    const items = fs.readdirSync(basePath);

    for (const item of items) {
        const itemPath = path.join(basePath, item);

        // Skip non-directories
        if (!fs.statSync(itemPath).isDirectory()) {
            continue;
        }

        // Skip excluded directories
        if (EXCLUDE.includes(item)) {
            continue;
        }

        // Skip hidden directories
        if (item.startsWith('.')) {
            continue;
        }

        // Check for .claude/skills directory
        const skillsPath = path.join(itemPath, '.claude', 'skills');
        if (!fs.existsSync(skillsPath) || !fs.statSync(skillsPath).isDirectory()) {
            continue;
        }

        // Scan skills directory
        const skills = [];
        const skillItems = fs.readdirSync(skillsPath);

        for (const skillItem of skillItems) {
            const skillItemPath = path.join(skillsPath, skillItem);
            const stat = fs.statSync(skillItemPath);

            if (stat.isDirectory()) {
                // Check if directory has SKILL.md
                const skillMdPath = path.join(skillItemPath, 'SKILL.md');
                if (fs.existsSync(skillMdPath)) {
                    skills.push({
                        name: skillItem,
                        type: 'skill',
                        path: skillItemPath
                    });
                } else {
                    skills.push({
                        name: skillItem,
                        type: 'dir_no_skill',
                        path: skillItemPath
                    });
                }
            } else if (stat.isFile() && path.extname(skillItem) === '.md') {
                // .md file directly in skills directory
                skills.push({
                    name: skillItem,
                    type: 'file',
                    path: skillItemPath
                });
            }
        }

        // Add project to results if it has skills
        if (skills.length > 0) {
            results[item] = {
                path: itemPath,
                skills_count: skills.filter(s => s.type === 'skill').length,
                items: skills
            };
        }
    }

    return results;
}

/**
 * Scan all configured directories for skills.
 *
 * @param {string[]} scanDirs - Directories to scan
 * @returns {Object} Results by scan directory
 */
function scanAllDirectories(scanDirs) {
    const allResults = {};

    for (const scanDir of scanDirs) {
        if (fs.existsSync(scanDir)) {
            allResults[scanDir] = findSkills(scanDir);
        } else {
            allResults[scanDir] = {};
        }
    }

    return allResults;
}

/**
 * Main execution function.
 */
async function main() {
    const allResults = scanAllDirectories(SCAN_DIRS);
    let totalSkills = 0;

    // Count total skills
    for (const scanDir in allResults) {
        const projects = allResults[scanDir];
        for (const projectName in projects) {
            totalSkills += projects[projectName].skills_count;
        }
    }

    // Print summary
    console.log('=== SCAN RESULTS ===\n');
    console.log(`Total skills found: ${totalSkills}\n`);

    for (const scanDir in allResults) {
        const projects = allResults[scanDir];
        if (Object.keys(projects).length > 0) {
            console.log(`\n--- ${scanDir} ---\n`);

            // Sort projects by name
            const sortedProjects = Object.entries(projects).sort((a, b) => a[0].localeCompare(b[0]));

            for (const [projectName, data] of sortedProjects) {
                const skillCount = data.skills_count;
                const otherCount = data.items.length - skillCount;

                let line = `${projectName}: ${skillCount} skills`;
                if (otherCount > 0) {
                    line += ` (+${otherCount} other files)`;
                }
                console.log(line);

                // Print items
                for (const item of data.items) {
                    const marker = item.type === 'skill' ? '[SKILL]' :
                                   item.type === 'file' ? '[FILE]' : '[DIR]';
                    console.log(`    ${marker} ${item.name}`);
                }
            }
        }
    }

    // Save JSON
    const outputPath = path.join(__dirname, '..', 'docs', 'skills-scan-results.json');
    await writeJson(outputPath, {
        total_skills: totalSkills,
        results: allResults
    });
    console.log(`\n\nJSON saved to: ${outputPath}`);
}

// CLI wrapper
/* istanbul ignore next */
if (require.main === module) {
    main().catch(error => {
        console.error('Error:', error.message);
        process.exit(1);
    });
}

module.exports = { findSkills, scanAllDirectories, main };
